plugin/cache: Fix prefetching issues

grobie opened the pull request:
### 1. What does this pull request do?

Several fixes around the prefetch behavior of plugin/cache. For more details, check the individual commit messages. The biggest behavioral changes are:

* new metric `coredns_cache_prefetch_total`
* allow prefetching of single requests
* allow prefetching of records with TTL < 10s (which was impossible before)
* prefetch in the background to not report wrong client request duration

### 2. Which issues (if any) are related?

#1278 

### 3. Which documentation changes (if any) need to be made?

The documentation of the available metrics and their labels has changed.

miekg commented on plugin/cache/cache.go:
```diff
@@ -109,8 +110,9 @@ func (w *ResponseWriter) WriteMsg(res *dns.Msg) error {
 
-		cacheSize.WithLabelValues(Success).Set(float64(w.pcache.Len()))
-		cacheSize.WithLabelValues(Denial).Set(float64(w.ncache.Len()))
+		l := strings.Join(w.Zones, ",")
```
This might be pretty expensive and create zone gc, is there a way around it?

miekg replied:
Why are we including this in the metrics?

grobie replied:
See the commit message and pr description.

grobie replied:
This can be written GC free by currying metrics. I'll look into it.

miekg commented on plugin/cache/handler.go:
```diff
@@ -37,25 +39,22 @@ func (c *Cache) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg)
-				i1.Freq.Reset(now, i.Freq.Hits())
+			threshold := int(math.Ceil(float64(c.percentage) / 100 * float64(i.origTTL)))
+			if i.Freq.Hits() > c.prefetch && ttl <= threshold {
+				go func() {
```
Why does this start another goroutine?

grobie replied:
See the commit message and pr description.

miekg commented:
Please add a test showing prefetch *isn't* working.

I don't understand why the zones are included like this, there is no per-zone partition in the cache.

grobie commented:
> I don't understand why the zones are included like this, there is no per-zone partition in the cache.

Ugh, you're right, gonna remove it again. Not being able to configure cache TTL per zone will make things a lot harder.

I find this mismatch between the config (where I can happily define multiple independent zones with separate settings) and the actual behavior (global plugin instances, every plugin has to check again for each request whether it matches) very confusing in CoreDNS. Is there a design document explaining the rationale for not isolating different chains from each other?

grobie commented:
> So TestPrefetch is unsufficient as well?

Actually, the existing TestPrefetch is broken, it doesn't test anything. You can remove the `p = true` line and it'll still pass.

grobie commented:
Alright, I refactored the tests to actually test the prefetch behavior and added one for the broken prefetching of responses with TTL < 10s. If you revert the code change of the percentage calculation to the original implementation, the test will fail now:

```
--- FAIL: TestPrefetch (1.00s)
        prefech_test.go:114: redis.service.dc1.consul.: want request to trigger a prefetch
        prefech_test.go:124: redis.service.dc1.consul.: want answer redis.service.dc1.consul.   4       IN      A       127.0.0.2, got redis.service.dc1.consul.        5       IN      A       127.0.0.2
```

grobie commented:
This PR depends on #1380 being merged first.

miekg commented on plugin/cache/prefech_test.go:
```diff
@@ -2,53 +2,181 @@ package cache
+			if want, got := 1, len(rec.Msg.Answer); want != got {
+				t.Errorf("%s: want %d answer RR, got %d", tt.qname, want, got)
+			}
+			if want, got := test.A(v.answer), rec.Msg.Answer[0]; !reflect.DeepEqual(want, got) {
```
Not a superfan of using reflect here, can you just compare the strings?

grobie replied:
Interesting, I've never heard an argument against reflect in test usage, especially DeepEqual is super common in all go code bases. But as you wish, will change it.

miekg commented on plugin/cache/handler.go:
```diff
@@ -25,37 +26,35 @@ func (c *Cache) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg)
-				i1.Freq.Reset(now, i.Freq.Hits())
+			threshold := int(math.Ceil(float64(c.percentage) / 100 * float64(i.origTTL)))
+			if i.Freq.Hits() >= c.prefetch && ttl <= threshold {
+				go func() {
```
(previous comment is gone?)

still don't see why this is needing to run in a goroutine?

grobie replied:
Is the commit message not helpful?

> The logging and metrics plugins (among others) included the duration of
a cache prefetch in the request latency of client request. This change
fixes this wrong reporting and executes the prefetch request in a
goroutine in the background.

The prefetch should not be counted towards the client request latency, as that can trip alerts on high latency.

grobie replied:
This doesn't matter. Other plugins like log and metrics wait for the termination of `ServeDNS()` (called by `NextOrFailure`) to determine the request latency.

miekg commented:
Test fails on master: (good!)

But it's quite elaborate (and from this output hard to debug). Can't we just have a simple prefetch/noprefetch (i.e. 2 or 3 test cases?)

~~~
--- FAIL: TestPrefetch (4.04s)
	prefech_test.go:153: prefetch.example.: want answer prefetch.example.	10	IN	A	127.0.0.1, got prefetch.example.	30	IN	A	127.0.0.1
	prefech_test.go:143: prefetch.example.: want request to trigger a prefetch
	prefech_test.go:153: prefetch.example.: want answer prefetch.example.	2	IN	A	127.0.0.1, got prefetch.example.	30	IN	A	127.0.0.1
	prefech_test.go:143: hits.reset.: want request to trigger a prefetch
	prefech_test.go:153: hits.reset.: want answer hits.reset.	7	IN	A	127.0.0.1, got hits.reset.	80	IN	A	127.0.0.1
	prefech_test.go:153: hits.reset.: want answer hits.reset.	73	IN	A	127.0.0.2, got hits.reset.	79	IN	A	127.0.0.1
	prefech_test.go:153: redis.service.dc1.consul.: want answer redis.service.dc1.consul.	4	IN	A	127.0.0.1, got redis.service.dc1.consul.	5	IN	A	127.0.0.1
	prefech_test.go:143: redis.service.dc1.consul.: want request to trigger a prefetch
	prefech_test.go:153: redis.service.dc1.consul.: want answer redis.service.dc1.consul.	1	IN	A	127.0.0.1, got redis.service.dc1.consul.	5	IN	A	127.0.0.1
	prefech_test.go:153: redis.service.dc1.consul.: want answer redis.service.dc1.consul.	4	IN	A	127.0.0.2, got redis.service.dc1.consul.	4	IN	A	127.0.0.1
	prefech_test.go:153: no.prefetch.: want answer no.prefetch.	15	IN	A	127.0.0.1, got no.prefetch.	30	IN	A	127.0.0.1
	prefech_test.go:153: no.prefetch.: want answer no.prefetch.	1	IN	A	127.0.0.1, got no.prefetch.	30	IN	A	127.0.0.1
	prefech_test.go:143: no.prefetch.: want request to trigger a prefetch
	prefech_test.go:153: no.prefetch.: want answer no.prefetch.	30	IN	A	127.0.0.2, got no.prefetch.	30	IN	A	127.0.0.1
FAIL
exit status 1
FAIL	github.com/coredns/coredns/plugin/cache	4.080s
~~~

grobie commented:
@miekg Aaaalright, I removed one test case and some verifications, and made the test output more readable.

```
--- FAIL: TestPrefetch (2.01s)
    --- FAIL: TestPrefetch/hits.reset. (1.00s)
        prefech_test.go:123: after 1m13s: want request to trigger a prefetch
    --- FAIL: TestPrefetch/short.ttl. (1.00s)
        prefech_test.go:123: after 4s: want request to trigger a prefetch
FAIL
FAIL    github.com/coredns/coredns/plugin/cache 2.018s
```

Given the different behaviors of the cache plugin (handler fetches from the next handler and returns that response, handler returns cached item and does not hit the next handler, handler returns cached item and does hit the next handler), I don't see how to remove any more test cases.

grobie commented on plugin/cache/cache.go:
```diff
@@ -32,6 +32,27 @@ type Cache struct {
+	now func() time.Time
+}
+
+// NewCache returns an initialized Cache with default settings. It's up to the
```
Done.

grobie commented on plugin/cache/prefech_test.go:
```diff
@@ -6,49 +6,158 @@ import (
+		verifications []verification
+	}{
+		{
+			qname:    "hits.reset.",
```
Done.

grobie commented on plugin/cache/setup.go:
```diff
@@ -115,6 +120,7 @@ func cacheParse(c *caddy.Controller) (*Cache, error) {
 					return nil, err
 				}
 				ca.ncap = ncap
+				ca.ncache = cache.New(ca.ncap)
```
My assumption was they'll get gargage collected and it doesn't matter. Without this change, two caches will always be initialized even if now explicit capacity attribute is set. I see the actual issue in only having a single cache instance and a single mutex which is shared across all zones. I changed it back.

miekg commented:
Thanks for pushing this!
